File: Data_Analysis/MotionCorrection.py
def MotionCorrection_smooth_yy(R,G,a0,rho_r,rho_g,mu_y,rho_y,mu_m,Um,Sminv,D):    
    """MotionCoRection_smooth_yy - estimate activity y(t) under GP prior on m(t)
       Computes MAP estimate of yy under the following model:
        mm ~ N(mu_m, Sigma_m)  # motion artifact
        yy ~ N(mu_y, rho_y*I)  # neural activity-relatd fluorescence
        aa = D\yy             # aa comes from yy via AR1 process 
        R = mm + noise        # measured rfp
        G = aa.*mm + noise    # measured gcamp
        Sigma_m = Ubasis*(sdqrt^2)*Ubasis' -- low-rank approximation to covariance

     Inputs:
        R      [Tx1]  - rfp measurements
        G      [Tx1]  - gcamp measurements
        a0     [Tx1]  - initial value for a(t)
        rho_r  [1]   - variance of rfp noise
        rho_g  [1]   - variance of gfp noise
        mu_y   [1]   - prior mean of y(t)
        rho_y  [1]   - prior variance of y(t)
        mu_m   [1]   - mean of motion artifact
        Ubasis [Txk]  - basis for rank-k approx to variance of motion artifact
        Sinv   [kxk]  - Sparse diagonal matrix with truncated inverse singular values of Cm
        D      [TxT]  - mapping from neural activity yy to activity-related fluorescence aa

     Output: 
         yy [Tx1] - estimate of neural activity"""
    
    lfunc = lambda y : Loss_MotionCoRection(y,R,G,rho_r,rho_g,mu_y,rho_y,mu_m,Um,Sminv,D)
    aa = minimize(lfunc,a0)
    return(aa.x)

def bsxfunmult (a,b):
    temp = np.zeros_like(a)
    for i in range(len(b)):
        temp[i,:] = a[i,:]*b[i]
    return(temp)

# ============ Loss function =====================================
def Loss_MotionCoRection(yy,R,G,rho_r,rho_g,mu_y,rho_y,mu_m,Um,Sminv,D):
    aa = np.linalg.solve(D,yy)
    
    
    # Log-determinant term
    dvec = (1/rho_r + 1./rho_g*(aa**2))

    M = Sminv + Um.T.dot(bsxfunmult(Um, dvec))
    (_,Mlogdet) = np.linalg.slogdet(M)
    trm_logdet = .5*(Mlogdet)


    # Diag term
    trm_diag = .5*(sum((R-mu_m)**2)/rho_r + sum((G-mu_m*aa)**2)/rho_g)

    # Quad term
    xt = Um.T.dot((R-mu_m)/rho_r + aa*(G-mu_m*aa)/rho_g)
    trm_quad = -.5*(xt.T.dot(np.linalg.solve(M,xt)))

    # Prior term
    trm_prior = .5/rho_y*(sum((yy-mu_y)**2))

    # Sum them up
    obj = trm_logdet + trm_diag + trm_quad + trm_prior
    return (obj)

File: MotionCorrection/MatlabCode/MotionCorrection.m
function y = MotionCorrection(rr,gg,y0,rho_r,rho_g,mu_y,rho_y,mu_m,Ubasis,sdiag,D)
% MOTIONCORRECTION - motion correction estimate under GP priors on m
%
% Computes MAP estimate of yy under the following model:
%    mm ~ N(mu_m, Sigma_m)  % motion artifact
%    yy ~ N(mu_y, rho_y*I)  % neural activity
%    aa = D\yy
%    rr = mm + noise        % measured rfp
%    gg = aa.*mm + noise    % measured gcamp
%    Sigma_m = Ubasis*sdiag*Ubasis' -- low-rank approximation to covariance
%
% Inputs:
%    rr     [Tx1]  - rfp measurements
%    gg     [Tx1]  - gcamp measurements
%    y0     [Tx1]  - initial value for neural activity estimate
%    rho_r  [1]   - variance of rfp noise
%    rho_g  [1]   - variance of gfp noise
%    mu_y   [1]   - prior mean of neural activity
%    rho_y  [1]   - prior variance of neural activity 
%    mu_m   [1]   - mean of motion artifact
%    Ubasis [Txk]  - basis for rank-k approx to variance of motion artifact
%    sdiag  [kxk]  - Diagonal matrix with truncated singular values of Sigma_m
%    D      [TxT]  - mapping from neural activity yy to activity-related fluorescence aa
%
% Output: 
%     yy [Tx1] - estimate of neural activity

lfunc = @(y) Loss_MotionCorrection(y,rr,gg,rho_r,rho_g,mu_y,rho_y,mu_m,Ubasis,sdiag,D);
opts = optimoptions('fminunc', 'display', 'iter', 'algorithm', 'quasi-newton');
y = fminunc(lfunc,y0,opts);
end

% ============ Loss function =====================================
function obj = Loss_MotionCorrection(y,rr,gg,rho_r,rho_g,mu_y,rho_y,mu_m,Ubasis,sdiag,D)

Tb = size(sdiag,1);

trm_prior =  1/(2*rho_y)*((y-mu_y)'*(y-mu_y)); % prior term

trm_logdet = 0.5*logdet(1/rho_r*sdiag + 1/rho_g*sqrt(sdiag)*Ubasis'*diag(D\y)*diag(D\y)*Ubasis*sqrt(sdiag) + eye(Tb)); % logdet term in posterior on yy

trm_quad= -0.5*(sqrt(sdiag)*Ubasis'*(rr./rho_r + diag(D\y)*gg./rho_g) + sqrt(sdiag)\Ubasis'*mu_m)'*...
    ((1/rho_r*sdiag + 1/rho_g*sqrt(sdiag)*Ubasis'*diag(D\y)*diag(D\y)*Ubasis*sqrt(sdiag) + eye(Tb))\eye(Tb))*...
    (sqrt(sdiag)*Ubasis'*(rr./rho_r + diag(D\y)*gg./rho_g) + sqrt(sdiag)\Ubasis'*mu_m); % likelihood term

obj = trm_prior + trm_quad + trm_logdet;
    
end